### DISCUSSION TITLE
Crud service fields encryption

### DISCUSSION THREAD

**Discussion by giovanni-motterle-mia-care** [mia-platform/community#541] [Q&A] [answered]
### What product are you having troubles with?

Marketplace

### What Console version are you using?

v13.0.5

### Description

Hi, I'm looking into the crud-service CSFLE feature.
I need to migrate encrypted data from one collection to another, accessing directly the DB.
I tried copying an object from one collection and inserting it into another. 
I was expecting the crud-service to not being able to decrypt the encrypted files of the copied object, but instead it is capable.
From my understanding, each collection has a dedicated data encryption key, how is it possible?

Thanks

**Comment by fredmaggiowski** [comment]
A few questions:

 - what MongoDB version (and license) are you using?
 - can you confirm data is actually encrypted in the original collection? (you should see a binary data instead of your plaintext value)
 - did the migration occur between two collections with the same name in different databases or two different collections in the same database?

**Reply by danibix95** [reply]
@giovanni-motterle-mia-care could you please also add the following details?
- CRUD Service version
- whether the two collections are managed or not by the same instance of CRUD Service

Indeed, in the former case, it is probably employing the same data key to access the data (more details on how the read/write mechanism works can be found [here](https://www.mongodb.com/docs/manual/core/csfle/fundamentals/automatic-encryption/#how-encrypted-writes-and-reads-work)

**Reply by giovanni-motterle-mia-care** [reply]
- I'm using mongoDB 6.0.16 (Atlas)
- yes, I can confirm that the field appears encrypted if I use compass to read the resource
- the migration occurred between 2 different collections in the same database (I tried also 2 different collections between 2 different databases (dev and preprod) and crud-service is unable to decrypt)
- I'm using crud-service 6.9.4
- the 2 collections are managed by the same crud-service instance

**Reply by danibix95** [reply, marked as answer]
I think I found the reason why CRUD Service is still able to decrypt the documents after being moved from a collection to another (managed by the same CRUD Service instance). As highlighted [in the MongoDB documentation](https://www.mongodb.com/docs/v6.1/core/csfle/reference/decryption/#metadata-used-for-decryption), when documents fields are encrypted, a set of metadata is added to the encrypted data, such as

- the unique id of the DEK employed to encrypt the document field
- the encryption algorithm used to encrypt the document field

Consequently, even though a DEK is created for each collection that contains documents with fields to be encrypted, a MongoDB client that has access to the same CMK and collection where the original DEKs were stored is able to decrypt all the documents that were previously encrypted with those keys, despite the documents may have moved from a collection to another.

**Reply by giovanni-motterle-mia-care** [reply]
Got it. Thanks!
